x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -792,6 +828,11 @@ static RealmRef newApiKeyRealmRef(String nodeName) {
             // no domain for API Key tokens
             return new RealmRef(API_KEY_REALM_NAME, API_KEY_REALM_TYPE, nodeName, null);
         }
+
+        static RealmRef newRemoteAccessRealmRef(String apiKeyId, String nodeName) {
+            // no domain for remote access authentication
+            return new RealmRef(REMOTE_ACCESS_REALM_NAME_PREFIX + "_" + apiKeyId, REMOTE_ACCESS_REALM_TYPE, nodeName, null);
```

Comment:
I made the suggestion of attaching API key ID to the realm name. But I am now questioning whether it is the right thing to do. It argubly makes the realm name look ugly. It leads to realm name explosion because each API key is now a diffferent realm. This could potentially cause issues for downstream processing, e.g. users want to aggregate information by realm name (the realm type is an alternative in some cases, but since you can have multiple realms of the same type, the alternative is not always viable).

I initially thought this could help ensure uniquness of users based on the underlying API keys. But now I think we don't really need it to achieve the same goal. Since we will have the API key ID in the Subject's metadata, uniqueness check should always be covered by, e.g:
* `canAccessResourceOf` - we will explicitly check the API key ID
* `equals`- this compares all metadata
* `hashCode` - this include all metadata as well

I think an analogy to this is the subject of an API key authentication. Without looking into the metadata for API key ID, two authentication of different API keys owned by the same user are essentially the same. But we do make sure checking into the metadata for uniqueness. So we don't have to rely on variation on the realm name. 

In summary, I now think we should just use a simple `_es_remote_access` for the realm name. What do you think? Sorry for the back and forth.

Reply:
My initial thought of making realm name unique by appending API key ID is because we have this notion that "a user is uniquely identifiable by its principal and realm". This is true for realm users within a single cluster and we rely on that for things like user profiles. Therefore I was trying to apply the same principle for the remote access users. But this notion does not really work outside of realm users (and their OAuth2 tokens). For example, it does not work for API keys. It certainly does not work for users from different clusters. Even if we suffix the realm name with API key ID, it is still possible that the same API key is used by two different QCs and the two QCs happen to have the exact same user/realm. They are hence indistinguishable to FC. 

When we are going to implement `canAccessResourcesOf` for the new remote access authentication, I think we will have to establish that the user identity uniqueness across different clusters relies on each cluster having a different API key. FC Admins need to create and distribute the API keys properly to ensure resources are not accidentally shared between users from different clusters.

Reply:
Thanks for thinking through this thoroughly, and apologies for not spotting this myself. I agree that `_es_remote_access_{ID}` may complicate things for end-users, e.g., if they want to agg by the `_es_remote_access` realm name. At the same time, they might actually want to agg by different API keys, that is aggregating by `_es_remote_access_{ID}` is exactly what they want. It's a little hard for me to tell, but I'm leaning towards your assessment that users will more often prefer to aggregate (or search) by `_es_remote_access`. That naming is also more in line with other synthetic realms, so I'm in favor of just `_es_remote_access` as the realm name.

> When we are going to implement canAccessResourcesOf for the new remote access authentication, I think we will have to establish that the user identity uniqueness across different clusters relies on each cluster having a different API key.

This is a really good point. My mental model here is that we shouldn't necessarily think of clusters as the authenticated entity, but the API key itself (or whatever other credential we might use in the future). Anything with access to the API key is treated as a single entity, from the point of view of the FC. So if we have two different querying clusters with users with overlapping principals they should be treated as one entity if the clusters use the same API key, and different entities if they use different keys. This is a potential gotcha for administrators and we'll have to be clear about this in the docs, but to me this is the only meaningful model for resource sharing. It's somewhat analogous to two different users sharing credentials to a single account -- there is nothing to be done about this, other than establishing that sharing passwords and other credentials is bad practice.